README.md
# AlphaBot2 - RobotsGo 

## Info
Repo containing all RobotsGo development code for the AlphaBot2 Platform  

### AlphaBot2-Control
Take control of the AlphaBot via gamepad or over a TCP/IP connection with keyboard while   
streaming the AB2 camera via ustreamer to webpage or player.       

For python3 only!!!!!

## Getting started
Set your AlphaBot2 up with Raspberry Pi OS https://www.raspberrypi.org/software/.   
Log in to your AlphaBot2 via your preferred method.      
Download the AlphaBot2-Setup.sh script make it executable and then run the script.   
```
$ cd ~
$ wget -L https://raw.githubusercontent.com/RobotsGo/AlphaBot2/main/AlphaBot2-Setup.sh
$ chmod +x AlphaBot2-Setup.sh
$ ./AlphaBot2-Setup.sh
```
This script will allow you to:    
* Update rpi os, install all needed dependencies, xrdp, clone RobotsGo/AlphaBot2 repo, enable ssh, enable required hardware support, update eeprom   
* Removes un-needed packages, setup Bluetooth XboxOne controller support, download and compile sixpair for PS3 Bluetooth pairing     
* Install all needed dependencies for client network control, clone RobotsGo's AlphaBot2 repo    
* Set the AlphaBot 2 up as a wifi 5Ghz AP with DHCP, WPA, SSID    

### Running AlphaBot2-Control

Launch the start script from ~/AlphaBot2/AlphaBot2-Control/   
```
$ ./start.sh 
```
From this script you can start:      
* ustreamer :Default port 8000    
* Network Controller client      
* Network Controller Server :Default port 5000        
* PS3 Controller Server    
* XboxOne Controller Server    
* MMP1251 Mod My PI Controller Server        
* ipega PG-9099 bluetooth Controller Server    
* View version info and RobotsGo links    

Streamer and Network Controller Server will auto set the IP of the AlphaBot     
When script exits the Streamer should stop

### Controller support   
PS3 Controller      
https://pimylifeup.com/raspberry-pi-playstation-controllers/        
R2 = Forwards,  L2 = Backwards, RightStick = Steering Left/Right         
LeftStick = Camera Up,Down,Left,Right,  Cross = Kicker            
Triangle = Center Camera, D UP/Down = Set Speed modes         
D Left/Right = Change led's colours, Square = Buzzer       
PS = Exit            

XboxOne Controller    
https://pimylifeup.com/xbox-controllers-raspberry-pi/      
xbox = Exit, RT = Forwards, LT = Backwards       
RightStick = Steering Left/Right, LeftStick = Camera Up,Down,Left,Right      
D UP/Down = Set Speed modes,  D Left/Right = Change led's colours      
A = Kick, Y = Center Camera, X = Buzzer       

Mod My Pi (mmp1251) Controller      
R2 = Forwards,  L2 = Backwards, HomeButton = exit          
RightStick = Steering Left/Right, A = kicker       
Y = Center Camera, X = Buzzer,  D UP/Down = Set Speed modes         
D Left/Right = Change led's colours, LeftStick = Camera Up,Down,Left,Right      

WaveShare rpi GameHat      
Start = exit,  Select = Center Camera,  TR = kick         
X = Forwards, B = Backwards,  Y = Left         
A = Right,  LeftStick = Camera Up,Down,Left,Right      

Keyboard - Network client   
Arrow Keys = Forwards, Backwards, Left, Right, Space = Kicker         
wasd = Camera Up,Down,Left,Right, q = Centre Camera         
1 2 3 = Speed Modes,   ,. = Change led's colours,  e = Buzzer     

ipega PG-9099 'WOLAVERINE' BlueTooth Controller   
Start = Exit, RT = Forwards, LT = Backwards          
RightStick = Steering Left/Right, LeftStick = Camera Up,Down,Left,Right         
D UP/Down = Set Speed modes,  D Left/Right = Change led's colours         
A = Kick, Y = Center Camera, X = Buzzer          


### TO DO
~~Add XboxOne controller support~~
~~Request Add MMP1251 Mod My Pi gamepad support~~   
~~Move as much functions over to the gpiozero library~~   
~~Use the buzzer~~            
~~Logo overlay on video stream~~   
Use the IR sensors   
Share data between the controller application and the streamer application (colour, speed modes)    
~~Get images and videos of unit in action~~       
~~Sort out rpi os install/script for GameHat + AB2~~
~~Improve camera streaming performance - investagate other ways~~ VLC requires 'Network Caching' to be reduced to 200ms to remove delay   
Build all this in to a wiki        


AlphaBot2-Control/ChangeLog.txt
***Version 0.1***19/05/2021
Major updates to setup script - nolonger need raspi-config, moved AP to 5ghz, added ustreamer
Removed flask and opencv replaced with ustreamer
Added ipega PG-9099 'WOLAVERINE' controler support
ustreamer in use for realtime http video streaming
code clean ups.

***Version 0.1***25/04/2021
Testing with setup script
Addded netifaces to mmp1251Mapping.py + networkMapping.py +
ps3Mapping.py + xboxOneMapping.py To help get ip when setup as AP
changed servo movment from 5 to 2 in servo.py

***Version 0.1***17/04/2021
small fixes to gameHatMapping.py + startGameHatClient.sh

***Version 0.1***16/04/2021
gameHatMapping.py - initial upload
startGameHatClient.sh - initial upload
networkMapping.py - fixes
streamer.py - new overlaty graphics **needs work**
Controllers.py - Added GameHat support

***Version 0.1*** 14/04/2021
Controllers.py + Gamepad.py pulled in to root dir, no longer need to clone gamepad repo
client.py - removed reference to version.py and add some more user information
mmp1251Mapping.py - initial upload
networkMapping.py - Added buzzer support
ps3Mapping.py - Code clean up's
start.sh - added new controller options XboxOne + mmp1251 (Mod My Pi) 
xboxOneMapping.py - initial upload

***Version 0.1*** 08/04/2021
Initial release


xboxOneMapping.py
#!/usr/bin/env python
#coding: utf-8

# Written by spoonie (Rick Spooner) for RobotsGo 

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import sys
import os
import Gamepad
import time
import led
import servo
import kick
import motors as motor
import socket
import buzzer
import netifaces
 
RUNNING = True
SPEED_MODE = 1 
SELECTED_COLOUR = 0  
COLOURS =  ['off', 'Red', 'Green', 'Blue', 'White']

def buzzerSoundOff():
    buzzer.buzzOff()

def buzzerSound():
    buzzer.buzzOn()

def centreServo():
    servo.centreServos()

def kickButtonPressed():
    kick.kickBall()

def selectSpeed(num):
    if num < 0:
        increaseSpeed()
    
    if num > 0:
        decreaseSpeed()

def selectColour(num):
    if num > 0:
        increaseColours()
    
    if num < 0:
        decreaseColours()

def increaseSpeed():
    global SPEED_MODE
    if SPEED_MODE < 3:
        SPEED_MODE = (SPEED_MODE + 1) 
        setSpeed()
        print('Speed mode ' + str(SPEED_MODE))
    
    else:
        print('Speed mode ' + str(SPEED_MODE))

def decreaseSpeed():
    global SPEED_MODE
    if SPEED_MODE > 1:
        SPEED_MODE = (SPEED_MODE - 1)
        setSpeed() 
        print('Speed mode ' + str(SPEED_MODE))
            
    else:
        print('Speed mode ' + str(SPEED_MODE))  

def increaseColours():
    global SELECTED_COLOUR
    global COLOURS
    if SELECTED_COLOUR < 4:
        SELECTED_COLOUR = (SELECTED_COLOUR + 1)
        led.setColour(SELECTED_COLOUR)
        print('Change colour to: ' + str(COLOURS[SELECTED_COLOUR]) + ' ' + str(SELECTED_COLOUR)) 
            
    else:
        print('Change colour to: ' + str(COLOURS[SELECTED_COLOUR]) + ' ' + str(SELECTED_COLOUR))

def decreaseColours():
    global SELECTED_COLOUR
    global COLOURS
    if SELECTED_COLOUR > 0:
        SELECTED_COLOUR = (SELECTED_COLOUR - 1)
        led.setColour(SELECTED_COLOUR) 
        print('Change colour to: ' + str(COLOURS[SELECTED_COLOUR]) + ' ' + str(SELECTED_COLOUR))
            
    else:
        print('Change colour to: ' + str(COLOURS[SELECTED_COLOUR]) + ' ' + str(SELECTED_COLOUR))

def setSpeed():
    #Speed tuning motor.Speed(MOTOR1, MOTOR2)
    #Looking from the front (USB PORTS) MOTOR1 is on your right
    #Increase the slower wheel 0.1 till bot moves stright ish. 
    if (SPEED_MODE == 1):
        motor.Speed(0.3, 0.2) #default (0.2, 0.2)

    elif(SPEED_MODE == 2):
        motor.Speed(0.6, 0.5) #default (0.5, 0.5)
        
    elif(SPEED_MODE == 3):
        motor.Speed(0.9, 0.8) #default (0.5, 0.5)
    
    else:
        motor.Speed(0.3, 0.2) #default (0.2, 0.2)
    
        
def exitButtonPressed():
    led.clearLed()
    centreServo()
    global RUNNING
    RUNNING = False
    sys.exit()

def xboxOneMappingStart():
    global SPEED_MODE
    SPEED_MODE = 1
    
    global SELECTED_COLOUR
    SELECTED_COLOUR = 0
    
    COLOURS =  ['off', 'Red', 'Green', 'Blue', 'White']
    
    centreServo()
    
    GAME_PAD_TYPE = Gamepad.XboxONE
    BUTTON_EXIT = 'HOME'
    TRIGGER_FORWARDS = 'RT'
    TRIGGER_BACKWARDS = 'LT'
    STEERING = 'RAS -X'
    CAMERA_VERTICAL = 'LAS -Y'
    CAMERA_HORIZONTAL = 'LAS -X'
    ADJUST_SPEED = 'DPAD -Y'
    CHANGE_COLOUR = 'DPAD -X'
    BUTTON_KICK = 'A'
    BUTTON_CENTRE_SERVOS = 'Y'
    BUTTON_BUZZER = 'X'
    POLL_INTERVAL = 0.1
    
    INT_FACE = netifaces.ifaddresses('wlan0')
    HOST_IP = INT_FACE[netifaces.AF_INET][0]['addr']
    
    print("AlphBot2 streaming and XboxOne control server version 0.1")
    print("RobotsGo web https://robotsgo.net/")
    print("RobotsGo webhttps://github.com/RobotsGo")
    print("")
    
    if HOST_IP != " ":
        print("Video Stream will be available @ http://" + str(HOST_IP) + ":8000")
        print("To use VLC etc, Video Stream will be available @ http://" + str(HOST_IP) + ":8000/stream")
    
    if not Gamepad.available():
        print('Please connect your gamepad...')
        while not Gamepad.available():
            time.sleep(1.0)
    gamepad = GAME_PAD_TYPE()
    
    #Buzz buzz when controller is connected.
    buzzerSound()
    time.sleep(0.1)
    buzzerSoundOff()
    time.sleep(0.1)
    buzzerSound()
    time.sleep(0.1)
    buzzerSoundOff()
    
    print('Gamepad connected, setup complete') 
    
    # Start the background updating
    gamepad.startBackgroundUpdates()
    gamepad.addButtonPressedHandler(BUTTON_KICK, kickButtonPressed)
    gamepad.addButtonPressedHandler(BUTTON_EXIT, exitButtonPressed)
    gamepad.addButtonPressedHandler(BUTTON_CENTRE_SERVOS, centreServo)
    gamepad.addButtonPressedHandler(BUTTON_BUZZER, buzzerSound)
    gamepad.addButtonReleasedHandler(BUTTON_BUZZER, buzzerSoundOff)
    gamepad.addAxisMovedHandler(ADJUST_SPEED, selectSpeed)
    gamepad.addAxisMovedHandler(CHANGE_COLOUR, selectColour)
    
    # Joystick events handled in the background
    try:
        while RUNNING and gamepad.isConnected():

            if gamepad.axis(TRIGGER_BACKWARDS) == 1:
                print('Backwards')
                motor.backwards()

            elif gamepad.axis(TRIGGER_FORWARDS) == 1:
                print('Forwards')
                motor.forwards()
            
            elif gamepad.axis(STEERING) == 1:
                print('Right')
                motor.right()

            elif gamepad.axis(STEERING) == -1:
                print('Left')
                motor.left()
            
            elif gamepad.axis(CAMERA_VERTICAL) == -1:
                servo.verticalUp()
                print('Camera UP ') 
            
            elif gamepad.axis(CAMERA_VERTICAL) == 1:
                servo.verticalDown()
                print('Camera Down ')
            
            elif gamepad.axis(CAMERA_HORIZONTAL) == -1:
                servo.horizontalLeft()
                print('Camera Left ')
                
            elif gamepad.axis(CAMERA_HORIZONTAL) == 1:
                servo.horizontalRight()
                print('Camera Right ')
        
            else:
                motor.stopMotor();
        
        time.sleep(POLL_INTERVAL)
        
    except Exception as e:
            print(e.message)
    finally:
    # Ensure the background thread is always terminated when we are done
        gamepad.disconnect()

xboxOneMappingStart()
